src/adapter/src/coord/read_policy.rs

```diff
@@ -110,9 +110,13 @@ impl<T: Eq + Hash + Ord> ReadHolds<T> {
     }
 
     /// Extends a `ReadHolds` with the contents of another `ReadHolds`.
-    pub fn extend(&mut self, other: ReadHolds<T>) {
+    /// Asserts that the newly added read holds don't coincide with any of the existing read holds in self.
+    pub fn extend_with_new(&mut self, other: ReadHolds<T>) {
         for (time, id_bundle) in other.holds {
-            self.holds.entry(time).or_default().extend(&id_bundle);
+            self.holds
+                .entry(time)
+                .or_default()
+                .extend_with_new(&id_bundle);
```

Comment:
I think we're better off doing something like
```Rust
pub fn extend_with_new(&mut self, other: ReadHolds<T>) {
        for (time, id_bundle) in other.holds {
            let my_id_bundle = self.holds
                .entry(time)
                .or_default();
            assert!(my_id_bundle.intersection(&id_bundle).is_empty(), "explanation that read holds can't track multiplicity and we'll leak a read hold");
            my_id_bundle.extend(&id_bundle);
        }
}
```

I'd prefer to limit the scope of the asserts to this method, so someone else doesn't accidentally hit them in the future. Then we can also add a meaningful message to the assert that explain exactly what went wrong and why it's wrong.

Reply:
Done: 95ef6894814e999b20dd5ef3743fa5232e5bb63d